### ISSUE TITLE
Windows Slave cannot connect since upgraded to Jenkins 2.150.3 and Plugin 1.2

### ISSUE THREAD

**Issue by mlueckert** [FarmGeek4Life/NegotiateSSO#12]
Dear All

Unfortunately I cannot get my Windows Slaves to connect after upgrading to Jenkins 2.150.3 and Plugin 1.2.

If I try to download the jnlp file via WebStart, I am getting the below stack trace:
```
java.lang.NullPointerException
	at jenkins.model.Jenkins.isSubjectToMandatoryReadPermissionCheck(Jenkins.java:4707)
	at com.github.farmgeek4life.jenkins.negotiatesso.NegSecFilter.shouldAttemptAuthentication(NegSecFilter.java:176)
	at com.github.farmgeek4life.jenkins.negotiatesso.NegSecFilter.doFilter(NegSecFilter.java:91)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at hudson.plugins.locale.LocaleFilter.doFilter(LocaleFilter.java:42)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.telemetry.impl.UserLanguages$AcceptLanguageFilter.doFilter(UserLanguages.java:128)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:157)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:105)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:135)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:90)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1340)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1242)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:503)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:364)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
	at java.lang.Thread.run(Unknown Source)
```

In the jenkins-slave.err.log of the slave I can see the following entries:
```
Failing to obtain http://xxx/computer/xxx/slave-agent.jnlp?encrypt=true
java.io.IOException: Failed to load http://xxx/computer/xxx/slave-agent.jnlp?encrypt=true: 500 Server Error
	at hudson.remoting.Launcher.parseJnlpArguments(Launcher.java:496)
	at hudson.remoting.Launcher.run(Launcher.java:322)
	at hudson.remoting.Launcher.main(Launcher.java:283)
Waiting 10 seconds before retry
```

As soon as I downgrade to plugin version 1.1 the slave comes online again, but this breaks SSO with the latest Jenkins version.

I also enabled Anonymous access to the slave but this did not work.

Thank you very much,
Marc

**Comment by FarmGeek4Life** [comment]
This is due to how the Jenkins code loads the encrypt parameter; it relies
on the current Stapler request, but NegotiateSSO operates on the request
before Stapler ever sees it. I'll probably have to re-duplicate some of the
Jenkins core code for dealing with this (I was hoping to get rid of that
code duplication...)

**Comment by mlueckert** [comment]
Thank you for the explanation.

So the only way to bring my Windows Slaves back online for now is to disable the NegotiateSSO.
Or do you see any other workaround?

Thanks, Marc

**Comment by FarmGeek4Life** [comment]
I have made changes that should fix this (checking for the slave-agent path before calling the Jenkins core method). Can you download the Snapshot build from https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fnegotiatesso-plugin/detail/master/5/artifacts and test it on your instance? If not, I can push this as a full release, I just can't test it because the instance of Jenkins I use doesn't have any slaves.

**Comment by mlueckert** [comment]
This looks good. I installed NegotiateSSO 1.3-SNAPSHOT (private-b74157c8-win2012-5bddc0$) from your link and SSO is still working and also the Windows slave now now connecting again.

**Comment by FarmGeek4Life** [comment]
Now released as NegotiateSSO 1.3.
